### ISSUE TITLE
expect_snapshot_value() fails with various errors with complex objects

### ISSUE THREAD

**Issue by DanChaltiel** [r-lib/testthat#1341]
My package is outputting MS Word documents and I've been wanting to test their reproducibility for ages. 

Third edition's snapshots are the perfect solution for that.

Obviouly, as a raw `docx` file is tracking a bunch of things that make it unreproducible (mostly dates and paths), `expect_snapshot_file()` will unfortunately not be usable in this case.

However, the `docx` object that allowed to save such a file is a plain R object that should be snapshotable.

Here is a reprex of comparability:

``` r
library(testthat)
library(officer)

#remove all non-reproducible internals
clean_docx = function(x){
    x$package_dir = "test"
    x$content_type$.__enclos_env__$private$filename = "test"
    x$doc_obj$.__enclos_env__$private$filename = "test"
    x$doc_obj$.__enclos_env__$private$package_dir = "test"
    x$doc_obj$.__enclos_env__$private$rels_filename = "test"
    x$footnotes$.__enclos_env__$private$filename = "test"
    x$footnotes$.__enclos_env__$private$package_dir = "test"
    x$footnotes$.__enclos_env__$private$rels_filename = "test"
    x
}

x=read_docx() %>% body_add_par("This is a test")
y=read_docx() %>% body_add_par("This is a test")

all.equal(x, y)
#> [1] "Component \"package_dir\": 1 string mismatch"                                                                                 
#> [2] "Component \"content_type\": Component \".__enclos_env__\": Component \"private\": Component \"filename\": 1 string mismatch"  
#> [3] "Component \"doc_obj\": Component \".__enclos_env__\": Component \"private\": Component \"filename\": 1 string mismatch"       
#> [4] "Component \"doc_obj\": Component \".__enclos_env__\": Component \"private\": Component \"package_dir\": 1 string mismatch"    
#> [5] "Component \"doc_obj\": Component \".__enclos_env__\": Component \"private\": Component \"rels_filename\": 1 string mismatch"  
#> [6] "Component \"footnotes\": Component \".__enclos_env__\": Component \"private\": Component \"filename\": 1 string mismatch"     
#> [7] "Component \"footnotes\": Component \".__enclos_env__\": Component \"private\": Component \"package_dir\": 1 string mismatch"  
#> [8] "Component \"footnotes\": Component \".__enclos_env__\": Component \"private\": Component \"rels_filename\": 1 string mismatch"
digest::digest(x)
#> [1] "95ae71952dcd2e3e7193785497523067"
digest::digest(y)
#> [1] "01f83d7fa2c6694f182dd537be5c9d6a"

x=read_docx() %>% body_add_par("This is a test") %>% clean_docx()
y=read_docx() %>% body_add_par("This is a test") %>% clean_docx()

all.equal(x, y)
#> [1] TRUE
digest::digest(x)
#> [1] "afb5e4f2db88b14df514097cdd4599a1"
digest::digest(y)
#> [1] "afb5e4f2db88b14df514097cdd4599a1"
```

<sup>Created on 2021-02-28 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

Unfortunately, this test (which should be reproducible) will fail with various error messages:

```r
library(officer)
test_that("word document", {
    local_edition(3)
    doc = read_docx() %>% 
        body_add_par("This is a test") %>% 
        clean_docx()
    
    expect_snapshot_value(doc, style = "json")
    #expect_snapshot_value(doc, style = "json2")
    #expect_snapshot_value(doc, style = "serialize")
})
```

With `json`, I get the error: "Error: No method asJSON S3 class: rdocx". (seems legit)  
With `json2`, I get the error: "Error: attempt to apply non-function".   
With `serialize`, I get the error: "Error: external pointer is not valid".

Also, would you consider allowing to use `digest::digest()` as an `expect_snapshot_value()` `style`? They are both unreadable but the former being shorter, it might be easier to compare and the output would be much less clumsy.   
Even better, you might want to consider letting the user choose any function, but that might require a hard work of abstraction that might not be relevant.

Side note, it seems that `serialize()` uses the R version ([link](https://stackoverflow.com/a/41876840/3888000)) which might lead to some unreproducibility if a file is compared to a snapshot that was created right before an R upgrade.

**Comment by DanChaltiel** [comment]
**TL;DR: this issue might be too complex to solve with reasonable effort, hence worth closing.**

On second thought, `rdocx` objects are so complex that my example was in fact wrong. Indeed, it ignored the actual content and two different documents ended having similar hashes or serializations:

``` r
library(officer)
library(testthat)
x=read_docx() %>% body_add_par("This is a test") %>% clean_docx()
y=read_docx() %>% body_add_par("This is a tedst") %>% clean_docx()

digest::digest(x)
#> [1] "8cdd37a720b41f7e7d643802418371fb"
digest::digest(y)
#> [1] "8cdd37a720b41f7e7d643802418371fb"
all.equal(
    serialize(x, NULL),
    serialize(y, NULL)
)
#> [1] TRUE
```
<sup>Created on 2021-02-28 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

It happened that I have to use `xml2::as_list()` to get inside the actual content of the document. 

Here is the actual code I'll be using. It allows me to snapshot the human-unreadable list, but also keep track of the actual documents, so I can compare them directly using MS Word:

```r
test_that("word document", {
    local_edition(3)
    doc = read_docx() %>% 
        body_add_par("This is a test")
       
    doc_xml = xml2::as_list(doc$doc_obj$get())
    
    x = expect_snapshot_value(doc_xml, style = "json2")
    
    filename="docx/4-officer/result_test_crosstable_officer.docx"
    if(!file.exists(filename)){
        dir.create("docx/4-officer/", recursive=TRUE, showWarnings=FALSE)
        print(doc, filename)
    }
    
    filename_new="docx/4-officer/result_test_crosstable_officer_new.docx"
    if(!inherits(x, "expectation_success")){
        print(doc, filename_new)
        warn(glue("Snapshot changed, see changes in files and :\n",
                  "   browseURL('tests/testthat/{filename}')\n",
                  "   browseURL('tests/testthat/{filename_new}')\n"))
    } else {
        if(file.exists(filename_new)) file.remove(filename_new)
        print(doc, filename)
    }       
})
```

**Comment by hadley** [comment]
Also note that you can always do `expect_snapshot_value(digest::digest(doc))`
